### PR TITLE
Using labels for container search and new listing functions

### LINKED ISSUE #536: collapse `eris ls`
**Problem** -- Users need to have a clear picture of `eris` at any one time. Basically my "dashboard" on the command line for what is happening with eris. We've built this functionality for the most part as `eris ls`. But what is outputted is (a) hard to quickly understand; (b) broken into too many pieces; (c) too long (unless I'm running an insanely large deployment it should not require me to scroll in my terminal window).

**Solution** -- Proposing a few specific UX changes to address this:

- [x] remove `--known` completely. As we gain in sophistication this is growing longer, but it isn't really helpful at all to users for what `eris ls` is trying to do. Indeed, I would advocate we should totally remove `--known` from all the listing functions but *only* expose them as `eris services ls --known`.... 
- [x] data containers which match to a service|chain container should not be displayed in `eris ls` (rather it should be a Yes|No in the output table immediately following `running` field). data containers which do not match to a service|chain container should be displayed
- [x] query... and i'm not sure. whether ports field is relevant (it probably is, but if folks feel strongly that it is not relevant information have no problem removing it as unnecessary cognitive load)
- [x] version information. `docker ps -a`'s image field is sometimes important for having an understanding (like is chainX running erisdb:0.11.2 or 0.11.3 or whatever) of what software the underlying binary for the service|chain is running. since eris' best practice is to utilize docker image tags here we should leverage this good practice to simplify the necessary information for users. i.e., we should display a `version` that maps to the docker image's tag. I would propose this field is placed between the `machine` and `running` fields in the current table.
- [x] remove actions from eris ls
- [x] vertical ordering should not be `chains` then `services` (as is current) but rather `services` then `chains` (with data only displayed if there are data containers which don't match to services/chains)
- [ ] add new field (the `autodocker` field) before container name. this will become more necessary as we have to plug more things "into" docker's daemon (e.g., obc, etc.). links to #472

### PR DESCRIPTION
**Pull request by pietv**
This PR implements the container searching functions using the information in container labels (closes the #258 issue) and simplifies the `eris ls` functions making them similar to `docker ps` (closes #536 and #483).

Fixed the issues with containers occasionally haven't been deleted with `eris clean` because they didn't have labels in them: in some places in the code the `DockerCreateData()` hasn't been prepended with the `LoadDataDefinition()` which sets the labels.

Fixed the `ls: cannot access /home/eris/.eris/chains/0C87B23A0DEF:` `No such file or directory`
`chains make` issue.

`util` directory file names have been rearranged so that most container related functions are now in 3 files: `containers.go`, `docker_client.go`, and `inspect.go`.

A utility `QueryYesOrNo` function for asking the `(y/n)` question. 

### Listing containers

* Standard output (`*` - is for running container, data section shows only containers for which there is no service or chain for). 

  ```
  $ eris services start keys
  $ eris services start ipfs
  $ eris chains new simple
  
  $ eris ls
    SERVICE       CONTAINER ID    DATA CONTAINER
  * ipfs          d7c674d588      c5f77812f9
  * keys          5c7bd26aaf      fb7627be77
                                                                                
    CHAIN         CONTAINER ID    DATA CONTAINER 
  * simple        925cdaf4ff      949728d0b7
                                                                                
    DATA          CONTAINER ID
  ```

* Extended output (adds `IMAGE`, `COMMAND`, and `PORTS` fields)

  ```
  $ eris ls -a
    SERVICE       CONTAINER ID    DATA CONTAINER   IMAGE                          COMMAND         PORTS
  * ipfs          d7c674d588      c5f77812f9       quay.io/eris/ipfs              [start]         0.0.0.0:4001->4001/tcp, 0.0.0.0:5001->5001/tcp, 0.0.0.0:8080->8080/tcp
                                                                                              
    CHAIN         CONTAINER ID    DATA CONTAINER   IMAGE                          COMMAND         PORTS
  * simple        925cdaf4ff      949728d0b7       quay.io/eris/erisdb:0.11.4     [new]           0.0.0.0:1337->1337/tcp, 0.0.0.0:46656->46656/tcp, 0.0.0.0:46657->46657/tcp
                                                                                              
    DATA          CONTAINER ID
    keys          fb7627be77

  ```
  
* Machine readable output (`--quiet`)
  
  ```
  $ eris services ls -q
  keys
  ipfs
  
  $ eris chains ls -q
  simple
  ```
* Machine readable output (`--json`)
  
  ```
  $ eris ls --json
  [
      {
        "Type": "service",
        "ShortName": "keys",
        "FullName": "ipfs-9f5324d4-3464-4908-a209-a4db29e8a0d9",
        "Labels": {
          "eris:ERIS": "true",
          "eris:NAME": "keys",
          "eris:TYPE": "service",
          "eris:USER": "Peter Vypov"
        },
        "Info": {
          "Id": "15d45c65ccdd097c53bcb69e22eeb60822e740b6192d28061becf926493a3a65",
          "Created": "2016-04-01T10:37:50.665411687Z",
          "Path": "eris-keys",
          ...
  ```
  
* Custom output (`--format`)

  ```
  $ eris ls -f '{{.ShortName}} {{.Type}} {{.Info.Path}}'
  keys service eris-keys
  ipfs service start
  simple chain erisdb-wrapper
  simple data true
  ipfs data true
  keys data true

  $ eris ls -f '{{.ShortName}}\t{{.Type}}\t{{.Info.NetworkSettings.IPAddress}}'
  keys            service         172.17.0.2
  ipfs            service         172.17.0.3
  simple          chain           172.17.0.4
  simple          data            
  ipfs            data            
  keys            data
  ```
  
### Listing known definition files

* Standard output

  ```
  $ eris services ls --known
  bigchaindb.toml      compilers.toml       keys.toml            openbazaar.toml      tor.toml 
  bitcoinclassic.toml  do_not_use.toml      logspout.toml        rethinkdb.toml       watchtower.toml 
  bitcoincore.toml     eth.toml             logsrotate.toml      tinydns.toml 
  btcd.toml            ipfs.toml            mindy.toml           toadserver.toml 
  
$ eris actions ls
  chain_info.toml    dns_register.toml  do_not_use.toml    keys_list.toml
  ```
* Custom output

  ```
  $ eris actions --known --format "{{.Name}}"
  chain_info
  dns_register
  do_not_use
  keys_list
  ```  

* Machine readable output (`--json`)
  
  ```
  $ eris chains ls --known --json
  [
      {
        "Name": "FEA1E40B78E5",
        "Definition": "/Users/peter/.eris/chains/FEA1E40B78E5.toml"
      },
      {
        "Name": "default",
        "Definition": "/Users/peter/.eris/chains/default.toml"
      },
      {
        "Name": "simple",
        "Definition": "/Users/peter/.eris/chains/simple.toml"
      }
  ]
  ```
  
### Container searching functions

Some container searching functions have been renamed or deleted:

Function|Description|Speed|Comments
--------|-----------|-----|--------
`UniqueName`|new UUID based container name|fast|
`ContainerName(type, name)`|Get a container name from type and short name|fast|Docker `ListContainers` if tne name is not cached
`Lookup(type, name)`|Get a container name from the cache|fast|
`ContainerDetails(long name)`|Get Docker Inspect details and Eris labels for the container|fast|Runs `Inspect` once
`ServiceContainerName(name)`|Get a long container name|fast|`ListContainers` if the name is not cached
`ChainContainerName(name)`|Get a long container name|fast|`ListContainers` if the name is not cached
`DataContainerName(name)`|Get a long container name|fast|`ListContainers` if the name is not cached
`ErisContainersByType(type, running)`|Discover containers of specific type|**slow**|`ListContainers` + `Inspect` for every container
`ErisContainers(filter(...){...}, running)`|Discover containers and filter them by their Docker `Inspect` or label values|**slow**|`ListContainers` + `Inspect` for every container
`IsService(name, running)`|Find a service|fast|`ListContainers`
`IsChain(name, running)`|Find a chain|fast|`ListContainers`
`IsData(name)`|Find a data|fast|`ListContainers`
`State(type, name, running)`|Find a container|fast|`ListContainers`
`Exists(type, name)`|Find an existing container|fast|`ListContainers`, former `HowManyContainersExisting`
`Running(type, name)`|Find a running container|fast|`ListContainers`, former `HowManyContainersRunning`

### CODE REVIEW

**Comment by pietv** [comment]
@compleatang, @zramsay please review. there are some pervasive changes and deletions.

**Comment by pietv** [comment]
Container IDs are used instead of shortened UUID container names because the latter are not copy-pastable into `docker rm`. (As an afterthought: 2 container names should do: one full UUID, and another truncated for display and `docker rm`)

File: chains/chains_test.go [hunk 1]
```diff
@@ -138,19 +99,19 @@ func TestStartChain(t *testing.T) {
 	defer tests.RemoveAllContainers()
 
 	start(t, chainName)
-	if n := util.HowManyContainersRunning(chainName, def.TypeChain); n != 1 {
-		t.Fatalf("expecting 1 chain container, got %v", n)
+	if !util.Running(def.TypeChain, chainName) {
```
Comment (compleatang) [inline chains/chains_test.go]: so much cleaner!

File: chains/load.go [hunk 2]
```diff
@@ -1,28 +0,0 @@
...
-
-func IsChainRunning(chain *definitions.Chain) bool {
-	log.WithField("=>", chain.Name).Debug("Checking chain running")
-	cName := util.FindChainContainer(chain.Name, false)
-	if cName == nil {
-		return false
-	}
-	chain.Operations.SrvContainerID = cName.ContainerID
-	return true
-}
```
Comment (compleatang) [inline chains/load.go]: :+1: that was an awful idea from the start!

File: chains/manage.go [hunk 3]
```diff
@@ -90,7 +92,7 @@ func MakeChain(do *definitions.Do) error {
 	doData := definitions.NowDo()
 	doData.Name = do.Name
 
-	doData.Operations.DataContainerName = util.DataContainersName(do.Name)
+	doData.Operations.DataContainerName = util.DataContainerName(do.Name)
```
Comment (compleatang) [inline chains/manage.go]: another super change. never should have had those functions so closely named....

File: cmd/list.go [hunk 4]
```diff
@@ -3,39 +3,77 @@ package commands
...
+    FullName  string          // container name
+
+    Labels map[string]string  // container labels
+    Info   *docker.Container  // Docker client library Container info 
+  }
+
+The full list of useful Container struct fields can be checked by issuing
+the [eris ls --json] command.
+
+Each field in the in the format input can be separated with the '\t' symbol
```
Comment (compleatang) [inline cmd/list.go]: `in the in the format`

File: cmd/list.go [hunk 5]
```diff
@@ -3,39 +3,77 @@ package commands
...
+
+  quote       quote the value
+  toupper     make the value upper case
+  ports       prettify exposed container ports (used as {{ports .Info}})
+  short       shorten the container ID (or any other value) to 10 symbols
+  asterisk    show the '*' symbol if the value is true
+  dependent   find a dependent data container for the given service or chain
+`,
+	Example: `$ eris ls -rf "{{.ShortName}}, {{.Type}}, {{ports .Info}}"
+$ eris ls  -f "{{.ShortName}}\t{{.Type}}\t{{.Info.NetworkSettings.IPAddress}}"`,
```
Comment (compleatang) [inline cmd/list.go]: :dart:

File: initialize/init.go [hunk 6]
```diff
@@ -116,23 +116,18 @@ func checkIfCanOverwrite(doYes bool) error {
...
 		"services path": common.ServicesPath,
 		"actions path":  common.ActionsPath,
 		"chains path":   common.ChainsPath,
 	}).Warn("Continuing may overwrite files in")
-	fmt.Print("Do you wish to continue? (y/n): ")
-	if _, err := fmt.Scanln(&input); err != nil {
-		return fmt.Errorf("Error reading from stdin: %v\n", err)
-	}
-	if input == "Y" || input == "y" || input == "YES" || input == "Yes" || input == "yes" {
+	if util.QueryYesOrNo("Do you wish to continue?") == util.Yes {
```
Comment (compleatang) [inline initialize/init.go]: :smoking:

File: chains/manage.go [hunk 7]
```diff
@@ -41,7 +41,8 @@ import (
 //  do.Debug         - debug output (optional)
 //
 func MakeChain(do *definitions.Do) error {
-	if err := checkKeysRunningOrStart(); err != nil {
+	err := checkKeysRunningOrStart()
+	if err != nil {
 		return err
 	}
```
Comment (zramsay) [inline chains/manage.go]: why lengthen this call?

File: list/container_stats.go [hunk 8]
```diff
@@ -1,171 +0,0 @@
...
-	Names := util.ContainerDisassemble(n)
-
-	v = Parts{
-		ShortName: Names.ShortName,
-		//Running: set in previous function
-		FullName:    Names.FullName,
-		PortsOutput: util.FormulatePortsOutput(contID),
-	}
-	return v, nil
-}
```
Comment (zramsay) [inline list/container_stats.go]: :) I hated this file so much

File: chains/manage.go (hunk 7 above)
Comment (pietv) [inline chains/manage.go]: You're right. No need. Probably for debugging statement in between. Will compress.

File: list/container_stats.go (hunk 8 above)
Comment (pietv) [inline list/container_stats.go]: Me too. :) Mostly because couldn't remember which functions in which file.

File: util/clean.go [hunk 9]
```diff
@@ -7,9 +7,10 @@ import (
 	"strings"
 
 	"github.com/eris-ltd/eris-cli/Godeps/_workspace/src/github.com/eris-ltd/common/go/common"
-	"github.com/eris-ltd/eris-cli/Godeps/_workspace/src/github.com/fsouza/go-dockerclient"
 
 	log "github.com/eris-ltd/eris-cli/Godeps/_workspace/src/github.com/Sirupsen/logrus"
+	docker "github.com/eris-ltd/eris-cli/Godeps/_workspace/src/github.com/fsouza/go-dockerclient"
+	def "github.com/eris-ltd/eris-cli/definitions"
```
Comment (zramsay) [inline util/clean.go]: imported packages from within cli should be above and seperate from anything in Godeps

File: util/clean.go [hunk 10]
```diff
@@ -166,19 +158,16 @@ func RemoveErisImages() error {
...
-		"images":     "All Eris docker images",
+		"containers": "all",
+		"scratch":    fmt.Sprintf("%s/.eris/scratch/data", home),
+		"rmd":        fmt.Sprintf("%s/.eris", home),
+		"images":     "all",
 	}
 
 	if toClean["all"] != true {
-		log.Warn("The marmots are about to remove")
+		log.Warn("The marmots are about to remove these Eris files")
```
Comment (zramsay) [inline util/clean.go]: it's not just files it might remove....

File: list/containers.go [hunk 11]
```diff
@@ -0,0 +1,246 @@
...
+		def.TypeData: {
+			Standard: {{t, false, dataTmplHeader, dataTmpl}},
+			Extended: {{t, false, dataTmplHeader, dataTmpl}},
+			Custom:   {{t, false, "", format}},
+		},
+		"all": {
+			Standard: {
+				{def.TypeService, false, standardTmplHeader, standardTmpl},
+				{def.TypeChain, false, standardTmplHeader, standardTmpl},
+				{def.TypeData, true, dataTmplHeader, dataTmpl},
```
Comment (compleatang) [inline list/containers.go]: am assuming vanilla `eris ls` resolves to roughly here. can we add a check in here (or elsewhere) for not adding the `DATA` section at all iff there are no hanging data containers?

i think that the DATA section will almost always be blank for folks and that that will be quite confusing as they wouldn't fully comprehend that we're just suppressing the container.

**Comment by zramsay** [comment]
++ LGTM. very excited to not be using `docker ps -a` anymore

File: util/clean.go (hunk 10 above)
Comment (pietv) [inline util/clean.go]: Yeah, I know. I thought about the term "objects" but didn't like it. The idea amending this line was to shorten the clean tags below, making them terser. Can't think of any better term. Maybe you have an idea?

**Comment by compleatang** [comment]
I second @zramsay ++ 

That is a huge amount of work! 

Other than the one comment on exploring suppressing that DATA section if no hanging data containers I'm :+1: LGTM

File: list/containers.go (hunk 11 above)
Comment (pietv) [inline list/containers.go]: Yep. Can amend it like that, np.

File: util/clean.go (hunk 10 above)
Comment (pietv) [inline util/clean.go]: Will change to something else.

Comment (compleatang) [inline util/clean.go]: can we just say everything?

Comment (compleatang) [inline util/clean.go]: `The marmots are about to remove the following` ...?

**Comment by zramsay** [comment]
i'm thinking the `*` should have a header (`STATUS` maybe?) ... there's a ~ amount of cognitive overhead otherwise. eg., the way docker-machine does it:
```
NAME      ACTIVE   DRIVER       STATE     URL                         
default   -        virtualbox   Stopped                                        
eris      *        virtualbox   Running   tcp://192.168.99.100:2376 
```
heck, it could even be `ON`, where the `*` implies running; or we have `-` to indicate off adding that latter option would make nicer on the eyes & will be easier to visually parse if ++ containers with a mix of running & not

**Comment by pietv** [comment]
Yep, sounds good :+1:  

New sample output:

```
$ eris services start ipfs
$ eris services start keys
$ eris services start tinydns
$ eris chains new simple

$ eris ls
SERVICE     ON     CONTAINER ID     DATA CONTAINER
tinydns     *      bf5b965c4a       8828d778af
keys        *      0d18d58c33       736ee8ec5f
ipfs        *      c4a9c7fc97       688361501d
                                                
CHAIN       ON     CONTAINER ID     DATA CONTAINER
simple      *      9ce905a1c7       6dc55f4326

$ docker rm -f 9ce905a1c7

$ eris ls
SERVICE     ON     CONTAINER ID     DATA CONTAINER
tinydns     *      bf5b965c4a       8828d778af
keys        *      0d18d58c33       736ee8ec5f
ipfs        *      c4a9c7fc97       688361501d
                                                
CHAIN       ON     CONTAINER ID     DATA CONTAINER
                                                
DATA        ON     CONTAINER ID
simple      -      6dc55f4326
```

**Comment by compleatang** [comment]
Merge away when you're happy with it good sir!

**Pull request merged**